lib/android.js

```diff
@@ -1498,54 +1659,534 @@ function unwrapMethodId (methodId) {
+    }
+
+    writer.putLabel('invoke_replacement');
+
+    writer.putLdrRegRegOffset('x16', 'x0', artMethodOffsets.quickCode);
+    writer.putBrReg('x16');
+
+    writer.flush();
+  });
+
+  return offset;
+}
+
+const artQuickCodePrologueWriters = {
+  ia32: writeArtQuickCodePrologueX86,
+  x64: writeArtQuickCodePrologueX86,
+  arm: writeArtQuickCodePrologueArm,
+  arm64: writeArtQuickCodePrologueArm64
+};
+
+function writeArtQuickCodePrologueX86 (target, trampoline) {
+  Memory.patchCode(target, 16, code => {
+    const writer = new X86Writer(code, { pc: target });
+
+    writer.putJmpAddress(trampoline);
+    writer.flush();
+  });
+}
+
+function writeArtQuickCodePrologueArm (target, trampoline) {
+  const targetAddress = target.and(THUMB_BIT_REMOVAL_MASK);
+
+  Memory.patchCode(targetAddress, 16, code => {
+    const writer = new ThumbWriter(code, { pc: targetAddress });
+
+    writer.putLdrRegAddress('pc', trampoline.or(1));
+    writer.flush();
+  });
+}
+
+function writeArtQuickCodePrologueArm64 (target, trampoline) {
+  Memory.patchCode(target, 16, code => {
+    const writer = new Arm64Writer(code, { pc: target });
+
+    writer.putBranchAddress(trampoline);
+    writer.flush();
+  });
+}
+
+class ArtQuickCodeInterceptor {
+  constructor (quickCode) {
+    this.quickCode = quickCode;
+    this.trampoline = null;
+    this.overwrittenPrologue = null;
+    this.overwrittenPrologueLength = 0;
+  }
+
+  _createTrampoline () {
+    if (trampolineAllocator === null) {
+      const trampolineSize = (pointerSize === 4) ? 128 : 256;
+      trampolineAllocator = makeCodeAllocator(trampolineSize);
+    }
+
+    this.trampoline = trampolineAllocator.allocateSlice();
+  }
+
+  activate (vm) {
+    this._createTrampoline();
+
+    const writeTrampoline = artQuickCodeReplacementTrampolineWriters[Process.arch];
+    const prologueLength = writeTrampoline(this.trampoline, this.quickCode, vm);
+    this.overwrittenPrologueLength = prologueLength;
+
+    const quickCodeAddress = (Process.arch === 'arm')
+      ? this.quickCode.and(THUMB_BIT_REMOVAL_MASK)
+      : this.quickCode;
+
+    this.overwrittenPrologue = Memory.dup(quickCodeAddress, prologueLength);
+
+    const writePrologue = artQuickCodePrologueWriters[Process.arch];
+    writePrologue(this.quickCode, this.trampoline);
+  }
+
+  deactivate () {
+    const quickCodeAddress = (Process.arch === 'arm')
+      ? this.quickCode.and(THUMB_BIT_REMOVAL_MASK)
+      : this.quickCode;
+
+    const prologueLength = this.overwrittenPrologueLength;
+
+    const Writer = thunkWriters[Process.arch];
+    Memory.patchCode(quickCodeAddress, prologueLength, code => {
+      const writer = new Writer(code, { pc: quickCodeAddress });
+
+      writer.putBytes(this.overwrittenPrologue.readByteArray(prologueLength));
+      writer.flush();
+    });
+
+    trampolineAllocator.freeSlice(this.trampoline);
+  }
+}
+
+function isArtQuickEntrypoint (address) {
+  const api = getApi();
+
+  const m = api.module;
```

Comment:
```suggestion
  const { module: m, artClassLinker: linker } = api.module;
```

^ perhaps something like this.